File: Difficulty: Medium/Count Inversions/count-inversions.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    // Function to count inversions in the array.
    int cnt = 0;
    void merge(vector<int> &arr, int l, int r){
        int mid = l + (r - l)/2;
        int n = mid - l + 1;
        int m = r - mid;
        
        vector<int> left(n);
        vector<int> right(m);
        
        int k = l;
        for(int i = 0; i<n; i++)
            left[i] = arr[k++];
        
        k = mid + 1;
        for(int i = 0; i < m; i++)
            right[i] = arr[k++];
        
        
        int i = 0, j = 0;
        k = l;
        while(i < n && j < m){
            if(left[i] <= right[j])
                arr[k++] = left[i++];
            else{
                // left > right
                arr[k++] = right[j++];
                cnt += (n-i);
            }
                
        }
        
        while(i < n)
            arr[k++] = left[i++];
        
        while(j < m)
            arr[k++] = right[j++];
    }
    
    void mergesort(vector<int> &arr, int l, int r){
        if(l >= r)
            return;
        
        int mid = l + (r - l)/2;
        mergesort(arr, l, mid);
        mergesort(arr, mid+1, r);
        merge(arr, l, r);
    }
    int inversionCount(vector<int> &arr) {
        // Your Code Here
        int n = arr.size();
        mergesort(arr, 0, n-1);
        return cnt;
    }
};


//{ Driver Code Starts.

int main() {

    int T;
    cin >> T;
    cin.ignore();
    while (T--) {
        int n;
        vector<int> a;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int num;
        while (ss >> num)
            a.push_back(num);
        Solution obj;
        cout << obj.inversionCount(a) << endl;
        cout << "~" << endl;
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Count Inversions/count-inversions.java
class Solution {
    //static int cnt;
    
    static int merge(int[] arr, int low, int mid, int high){
        int n = mid - low + 1;
        int m = high - mid;
        
        int cnt = 0;
        
        int[] left = new int[n];
        int[] right = new int[m];
        
        for(int i = 0; i < n; i++)
            left[i] = arr[low + i];
            
        for(int j = 0; j < m; j++)
            right[j] = arr[mid + 1 + j];
            
        int i = 0, j = 0, k = low;
        while(i < n && j < m){
            if(left[i] <= right[j])
                arr[k++] = left[i++];
            else{
                cnt += n - i;
                arr[k++] = right[j++];
            }
        }
        
        while(i < n)
            arr[k++] = left[i++];
        while(j < m)
            arr[k++] = right[j++];
        
        return cnt;
        
    }
    
    static int mergeSort(int[] arr, int low, int high){
        if(low >= high)
            return 0;
        
        int cnt = 0;
        
        int mid = low + (high - low)/2;
        cnt += mergeSort(arr, low, mid);
        cnt += mergeSort(arr, mid+1, high);
        cnt += merge(arr, low, mid, high);
        
        return cnt;
    }
    static int inversionCount(int arr[]) {
        // Code Here
        int low = 0, high = arr.length - 1;
        int cnt = mergeSort(arr, low, high);
        return cnt;
    }
}